backend/modules/model_gateway/model_gateway.py

```diff
@@ -1,85 +1,119 @@
+                        logger.info(f"Adding model {model_name} to the model gateway.")
+                        self.modelsToProviderMap[model_name] = list
+                    else:
+                        # Avoid adding the model to the map if the API key is not set
+                        logger.warning(
+                            f"Environment variable {list.api_key_env_var} not set."
+                        )
                 for model_id in list.llm_model_ids:
-                    model_name = f"{list.provider_name}/{model_id}"
-                    self.modelsToProviderMap[model_name] = list
-            
+                    if list.api_key_env_var and os.environ.get(list.api_key_env_var):
+                        # Register only those models whose API key is set
+                        model_name = f"{list.provider_name}/{model_id}"
+                        logger.info(f"Adding model {model_name} to the model gateway.")
+                        self.modelsToProviderMap[model_name] = list
+                    else:
+                        # Avoid adding the model to the map if the API key is not set
+                        logger.warning(
+                            f"Environment variable {list.api_key_env_var} not set."
+                        )
+
     def get_embedding_models(self) -> List[ModelConfig]:
         models: List[ModelConfig] = []
         for list in self.config:
-            for model_id in list.embedding_model_ids:
-                models.append(ModelConfig(
-                    name=f"{list.provider_name}/{model_id}", 
-                    type=ModelType.embedding
-                ).dict())
-                              
+            if list.api_key_env_var and os.environ.get(list.api_key_env_var):
+                # Register only those models whose API key is set
+                for model_id in list.embedding_model_ids:
+                    models.append(
+                        ModelConfig(
+                            name=f"{list.provider_name}/{model_id}",
+                            type=ModelType.embedding,
+                        ).dict()
+                    )
+
         return models
-    
+
     def get_llm_models(self) -> List[ModelConfig]:
         models: List[ModelConfig] = []
         for list in self.config:
-            for model_id in list.llm_model_ids:
-                models.append(ModelConfig(
-                    name=f"{list.provider_name}/{model_id}", 
-                    type=ModelType.chat
-                ).dict())
-                              
+            if list.api_key_env_var and os.environ.get(list.api_key_env_var):
+                # Register only those models whose API key is set
+                for model_id in list.llm_model_ids:
+                    models.append(
+                        ModelConfig(
+                            name=f"{list.provider_name}/{model_id}", type=ModelType.chat
+                        ).dict()
```

Comment:
Why `dict` ?
the return type says `List[ModelConfig]`

Reply:
it was leading to JSON serialization error. Added a `to_dict()` to pydantic